services/databaseEncryptionService.ts

```diff
@@ -0,0 +1,348 @@
+import * as SecureStore from 'expo-secure-store';
+import * as Crypto from 'expo-crypto';
+import { gcm } from '@noble/ciphers/aes.js';
+
+const SECURE_STORE_KEYS = {
+  ENCRYPTED_DEK: 'encrypted_dek',
+  KEK: 'kek',
+  ENCRYPTION_MODE: 'encryption_mode',
+};
+
+export type EncryptionMode = 'basic' | 'protected';
+
+export class EncryptionError extends Error {
+  constructor(public code: string, message: string) {
+    super(message);
+    this.name = 'EncryptionError';
+  }
+}
+
+export const ERROR_CODES = {
+  USER_CANCELLED: 'USER_CANCELLED',
+  KEY_CORRUPTION: 'KEY_CORRUPTION',
+  SECURE_STORE_UNAVAILABLE: 'SECURE_STORE_UNAVAILABLE',
+  REWRAP_FAILED: 'REWRAP_FAILED',
+  AUTH_IN_PROGRESS: 'AUTH_IN_PROGRESS',
+} as const;
+
+let dekCache: string | null = null;
+
+function generateRandomKey(): string {
+  const randomBytes = Crypto.getRandomBytes(32);
+  return Array.from(randomBytes)
+    .map(b => b.toString(16).padStart(2, '0'))
+    .join('');
+}
+
+function hexToBytes(hex: string): Uint8Array {
+  const bytes = new Uint8Array(hex.length / 2);
+  for (let i = 0; i < hex.length; i += 2) {
+    bytes[i / 2] = parseInt(hex.substring(i, i + 2), 16);
+  }
+  return bytes;
+}
+
+function bytesToHex(bytes: Uint8Array): string {
+  return Array.from(bytes)
+    .map(b => b.toString(16).padStart(2, '0'))
+    .join('');
+}
+
+function stringToBase64(str: string): string {
+  const bytes = new TextEncoder().encode(str);
+  let binary = '';
+  for (let i = 0; i < bytes.length; i++) {
+    binary += String.fromCharCode(bytes[i]);
+  }
+  return btoa(binary);
+}
+
+function base64ToString(base64: string): string {
+  const binary = atob(base64);
+  const bytes = new Uint8Array(binary.length);
+  for (let i = 0; i < binary.length; i++) {
+    bytes[i] = binary.charCodeAt(i);
+  }
+  return new TextDecoder().decode(bytes);
+}
```

Comment:
Are you sure you do not have a built-in way to do base 64 encoding? In Node.js you'd use `Buffer.from()` and `buffer.toString()` - isn't there something like that available to you?